---
title: "Introduction to Splatter"
author: "Luke Zappia"
date: "`r Sys.Date()`"
output:
    BiocStyle::html_document:
        toc: true
vignette: >
  %\VignetteIndexEntry{An introduction to the Splatter package}
  %\VignetteEngine{knitr::rmarkdown}
  \usepackage[utf8]{inputenc}
---

```{r knitr-options, echo = FALSE, message = FALSE, warning = FALSE}
# To render an HTML version that works nicely with github and web pages, do:
# rmarkdown::render("vignettes/splatter.Rmd", "all")
knitr::opts_chunk$set(fig.align = 'center', fig.width = 6, fig.height = 5,
                      dev = 'png')
```

Welcome to Splatter! Splatter is an R package for the simple simulation of
single-cell RNA-seq data. This vignette gives an overview and introduction to
Splatter's functionality.

# Quickstart

Assuming you already have a matrix of count data similar to that you wish to
simulate there are two simple steps to creating a simulated data set with
Splatter. Here is an example using the example dataset in the `scater` package:

```{r quickstart}
library(splatter)

data("sc_example_counts")
params <- splatEstimate(sc_example_counts)
sim <- splatSimulate(params, dropout.present = FALSE)
```

These steps will be explained in detail in the following sections but briefly
the first step takes a dataset and estimates simulation parameters from it and
the second step takes those parameters and simulates a new dataset.

# The Splatter model

Before we look at how we estimate parameters let's first look at how Splatter
simulates data and what those parameters are. The core of the Splatter model is
a gamma-Poisson distribution used to generate a gene by cell matrix of counts.
Mean expression levels for each gene are simulated from a gamma distribution
and the Biological Coefficient of Variation is used to enforce a mean-variance
trend before counts are simuated from a Poisson distribution. Splatter also
allows you to simulate expression outlier genes (genes with mean expression
outside the gamma distribution) and dropout (random knock out of counts based
on mean expression). Each cell is given an expected library size (simulated
from a log-normal distribution) that makes it easier to match to a given
dataset.

Splatter can also simulate differential epression between groups of cells or
differentiation paths between different cells types. These are described further
in the [simulating counts] section.

## Parameters

The parameters required by Splatter are briefly described here:

* **Global parameters**
    * `nGenes` - The number of genes to simulate.
    * `nCells` - The number of cells to simulate.
    * `nGroups` - The number of groups or paths to simulate.
    * `groupCells` - The number of cells in each group/path.
    * `seed` - Seed to use for generating random numbers.
* **Mean parameters**
    * `mean.shape` - Shape parameter for the mean gamma distribution.
    * `mean.rate` - Rate parameter for the mean gamma distribution.
* **Library size parameters**
    * `lib.loc` - Location (meanlog) parameter for the library size log-normal
      distribution.
    * `lib.scale` - Scale (sdlog) parameter for the library size log-normal
      distribution.
* **Expression outlier parameters**
    * `out.prob` - Probability that a gene is an expression outlier.
    * `out.loProb` - Probability that an expression outlier gene is lowly
      expressed.
    * `out.facLoc` - Location (meanlog) parameter for the expression outlier
      factor log-normal distribution.
    * `out.facScale` - Scale (sdlog) parameter for the expression outlier factor
      log-normal distribution.
* **Differential expression parameters**
    * `de.prob` - Probability that a gene is differentially expressed in each
      group or path.
    * `de.loProb` - Probability that a differentially expressed gene is
      down-regulated.
    * `de.facLoc` - Location (meanlog) parameter for the differential expression
      factor log-normal distribution.
    * `de.facScale` - Scale (sdlog) parameter for the differential expression
      factor log-normal distribution.
* **Biological Coefficient of Variation parameters**
    * `bcv.common` - Underlying common dispersion across all genes.
    * `bcv.df` - Degrees of Freedom for the BCV inverse chi-squared
      distribution.
* **Dropout parameters**
    * `dropout.present` - Logical. Whether to simulate dropout.
    * `dropout.mid` - Midpoint parameter for the dropout logistic function.
    * `dropout.shape` - Shape parameter for the dropout logistic function.
* **Differentiation path parameters**
    * `path.from` - Vector giving the originating point of each path.
    * `path.length` - Vector giving the number of steps to simulate along each
      path.
    * `path.skew` - Vector giving the skew of each path.
    * `path.nonlinearProb` - Probability that a gene changes expression in a
      non-linear way along the differentiation path.
    * `path.sigmaFac` - Sigma factor for non-linear gene paths.

While this may look like a lot of parameters Splatter attempts to make it easy
for the user, both by providing sensible defaults and making it easy to estimate
many of the parameters from real data. For more details on the parameters see
`?SplatParams`.

# The `SplatParams` object

All the parameters for the Splatter simulations are stored in a `SplatParams`
object. Let's create a new one and see what it looks like.

```{r SplatParams}
params <- newSplatParams()
params
```

As well as telling us what type of object we have ("A `Params` object of class
`SplatParams`") and showing us the values of the parameter this output gives us
some extra information. We can see which parameters can be estimated by the
`splatEstimate` function (those in parentheses), which can't be estimated
(those in brackets) and which have been changed from their default values (those
in ALL CAPS). 

## Getting and setting

If we want to look at a particular parameter, for example the number of genes to
simulate, we can extract it using the `getParam` function:

```{r getParam}
getParam(params, "nGenes")
```

Alternatively to give a parameter a new value we can use the `setParam`
function:

```{r setParam}
params <- setParam(params, "nGenes", 5000)
params
```

The `nGenes` parameter is now shown in ALL CAPS to indicate it has been changed
form the default.

If we want to extract multiple parameters (as a list) or set multiple parameters
we can use the `getParams` or `setParams` functions:

```{r getParams-setParams}
# Extract multiple parameters as a list
getParams(params, c("nGenes", "mean.rate", "mean.shape"))
# Set multiple parameters at once (using a list)
params <- setParams(params, update = list(nGenes = 8000, mean.rate = 0.5))
params
# Set multiple parameters at once (using additional arguments)
params <- setParams(params, mean.shape = 0.5, de.prob = 0.2)
params
```

We can also set parameters directly when we call `newSplatParams`:

```{r newSplatParams-set}
params <- newSplatParams(lib.loc = 12, lib.scale = 0.6)
params
```

# Estimating parameters

Splatter allows you to estimate many of it's parameters from a data set
containing counts using the `splatEstimate` function.

```{r splatEstimate}
# Check that sc_example counts is an integer matrix
class(sc_example_counts)
typeof(sc_example_counts)
# Check the dimensions, each row is a gene, each column is a cell
dim(sc_example_counts)
# Show the first few entries
sc_example_counts[1:5, 1:5]

params <- splatEstimate(sc_example_counts)
```

Here we estimated parameters from a counts matrix but `splatEstimate` can also
take an `SCESet` object from the `scater` package. The estimation process has
the following steps:

1. Mean parameters are estimated by fitting a gamma distribution to the mean
   expression levels.
2. Library size parameters are estimated by fitting a log-normal distribution to
   the library sizes.
3. Expression outlier parameters are estimated by determining the number of
   outliers and fitting a log-normal distribution to their difference from the
   median.
4. BCV parameters are estimated using the `estimateDisp` function from the
   `edgeR` package.
5. Dropout parameter are estimated by checking if dropout is present and fitting
   a logistic function to the relationship between mean expression and
   proportion of zeros.
   
For more details of the estimation procedures see `?splatEstimate`.

# Simulating counts

Once we have a set of parameters we are happy with we can use `splatSimulate`
to simulate counts. If we want to make small adjustments to the parameters we
can provide them as additional arguments, alternatively if we don't supply any
parameters the defaults will be used:

```{r splatSimulate}
sim <- splatSimulate(params, nGenes = 1000, dropout.present = FALSE)
sim
```

Looking at the output of `splatSimulate` we can see that `sim` is an `SCESet`
object with `r dim(sim)[1]` features (genes) and `r dim(sim)[2]` samples
(cells). An `SCESet` combines a counts matrix (and other expression measures)
with phenotype information about each cell (accessed using `pData`) and feature
information about each gene (accessed using `fData`). Splatter uses these slots
to store information about the intermediate values of the simulation.

```{r SCESet}
# Information about genes
head(fData(sim))
# Information about cells
head(pData(sim))
# Gene by cell matrices
names(assayData(sim))
# Example of cell means matrix
assayData(sim)$CellMeans[1:5, 1:5]
# Access the counts
counts(sim)[1:5, 1:5]
```

An additional (big) advantage of outputting an `SCESet` is that we get immediate
access to all of the `scater` functions. For example we can make a PCA plot:

```{r pca}
plotPCA(sim)
```

For more details of the `SCESet` and what you can do with `scater` refer to the
`scater` documentation and [vignette][scater-vignette].

The `splatSimulate` function outputs the following additional information about
the simulation:

* **Cell information (`pData`)**
    * `Cell` - Unique cell identifier.
    * `Group` - The group or path the cell belongs to.
    * `ExpLibSize` - The expected library size for that cell.}
    * `Step` (paths only) - How far along the path each cell is.
* **Gene information (`fData`)**
    * `Gene` - Unique gene identifier.
    * `BaseGeneMean` - The base expression level for that gene.
    * `OutlierFactor` - Expression outlier factor for that gene.
    * `GeneMean` - Expression level after applying outlier factors.
    * `DEFac[Group]` - The differential expression factor for each gene
      in a particular group.
    * `GeneMean[Group]` - Expression level of a gene in a particular group after
      applying differential expression factors.
* **Gene x cell information (`assayData`)**
    * `BaseCellMeans` - The expression of genes in each cell adjusted for
      expected library size.
    * `BCV` - The Biological Coefficient of Variation for each gene in
      each cell.
    * `CellMeans` - The expression level of genes in each cell adjusted
      for BCV.
    * `TrueCounts` - The simulated counts before dropout.
    * `Dropout` - Logical matrix showing which counts have been dropped in which
      cells.

Values that have been added by Splatter are named using `UpperCamelCase` to
separate them from the `underscore_naming` used by `scater`. For more
information on the simulation see `?splatSimulate`.

## Simulating groups

So far we have only simulated a single population of cells but often we are
interested in investigating multiple a mixed population of cells and looking
to see what cell types are present or what differences there are between them.
Splatter is also able to simulate this situation by changing the method
argument Here we are going to simulate two groups, each with 50 cells, by
specifying the `groupCells` parameter and setting the `method` parameter to
`groups`:

```{r groups}
sim.groups <- splatSimulate(groupCells = c(100, 100), method = "groups",
                            verbose = FALSE)
plotPCA(sim.groups, colour_by = "Group")
```

## Simulating paths

The other situation that is often of interest is a differentiation process where
one cell type is changing into another. Splatter appoximates this process by
simulating a series of steps between one two groups and randomly assigning each
cell to a step. We can create this kind of simulation using the `paths` method.

```{r paths}
sim.paths <- splatSimulate(method = "paths", verbose = FALSE)
plotPCA(sim.paths, colour_by = "Step")
```

## Convenience functions

Each of the Splatter simulation methods has it's own convenience function.
To simulate a single population use `splatSimulateSingle()` (equivalent to
`splatSimulate(method = "single")`), to simulate grops use
`splatSimulateGroups()` (equivalent to `splatSimulate(method = "groups")`) or to
simulate paths use `splatSimulatePaths()` (equivalent to
`splatSimulate(method = "paths")`).

# Other simulations

```{r sessionInfo}
sessionInfo()
```

[scater-vignette]: https://bioconductor.org/packages/release/bioc/vignettes/scater/inst/doc/vignette.html